### ISSUE TITLE
Logging can cause errors or unresponsive services

### ISSUE THREAD

**Issue by hashstat** [VOLTTRON/volttron#108]
Logging handlers are currently created in the main thread using file objects that are not gevent-aware and log records can be emitted from both gevent and non-gevent contexts/threads. This can cause gevent threadlets to become unresponsive if a write to the log blocks. Wrapping the log file object in a gevent.fileobject.FileObject will fix the problem for gevent threadlets but will raise an exception in other threads if writing would block.

One possible solution is to write a smart handler or file object which is aware of the context and acts appropriately for that context.

**Comment by craig8** [comment]
@kmonson want to take this one on/determine if necessary?

**Comment by kmonson** [comment]
sure

**Comment by kmonson** [comment]
Have we every actually observed this behavior?

**Comment by acedrew** [comment]
@kmonson
Yes, we can definitely say we've observed this behavior now 😁 
Love the fact there was an existing issue for something that took me a few days of effort to debug... 🤕

**Comment by jhaack** [comment]
Can you give steps for replicating?

**Comment by acedrew** [comment]
Yes, point your logs at an rsyslog daemon, then kill rsyslog (may take some work to make it stay dead)
Where this was showing up in production was with verbose (DEBUG) logging turned on, and rsyslog shipping logs over the network. Slow connectivity may have exacerbated it, eventually rsyslog buffers would fill and it would begin to affect VOLTTRON

**Comment by jhaack** [comment]
@kmonson  take a look at this in the current context
